### PR TITLE
Match forward headers case insensitively.

### PR DESCRIPTION
**Pull request by yinggeh**
Currently, headers are lowercased before they are sent to the server, which causes inconvenience. According to HTTP specification header names are case-insensitive. Therefore, the forward header pattern (regex) should match the header case insensitively.

### CODE REVIEW

File: qa/L0_parameters/parameters_test.py [hunk 1]
```diff
@@ -54,13 +54,14 @@ async def asyncSetUp(self):
...
 
         self.parameter_list = []
-        self.parameter_list.append({"key1": "value1", "key2": "value2"})
-        self.parameter_list.append({"key1": 1, "key2": 2})
-        self.parameter_list.append({"key1": 123.123, "key2": 321.321})
-        self.parameter_list.append({"key1": True, "key2": "value2"})
-        self.parameter_list.append({"triton_": True, "key2": "value2"})
-
-        if TEST_HEADER == "1":
+        if TEST_HEADER in ["0","1"]:
```
Comment (rmccorm4) [inline qa/L0_parameters/parameters_test.py:57]: Why is this part now included for the `TEST_HEADER=0` case?

File: src/command_line_parser.cc [hunk 2]
```diff
@@ -1235,6 +1235,11 @@ TritonParser::Parse(int argc, char** argv)
   triton::server::grpc::Options& lgrpc_options = lparams.grpc_options_;
 #endif  // TRITON_ENABLE_GRPC
 
+#if defined TRITON_ENABLE_HTTP || defined TRITON_ENABLE_GRPC
+  // According to HTTP specification header names are case-insensitive.
+  const std::string case_insensitive_prefix{"(?i)"};
```
Comment (rmccorm4) [inline src/command_line_parser.cc:1240]: Looks clean 🚀 I think this approach will also let users disable the case insensitivity if they **really** need to for some unforeseen reason. ex: `--header-pattern "(?-i)..."` to disable it, which I think is a good flexibility to have.

File: qa/L0_parameters/parameters_test.py [hunk 3]
```diff
@@ -54,13 +54,14 @@ async def asyncSetUp(self):
...
+            self.parameter_list.append({"key1": "value1", "key2": "value2"})
+            self.parameter_list.append({"key1": 1, "key2": 2})
+            self.parameter_list.append({"key1": 123.123, "key2": 321.321})
+            self.parameter_list.append({"key1": True, "key2": "value2"})
+            self.parameter_list.append({"triton_": True, "key2": "value2"})
+
+        if TEST_HEADER in ["1","2"]:
```
Comment (rmccorm4) [inline qa/L0_parameters/parameters_test.py:64]: Need to run the pre-commit hooks to fix the linter complaints in the PR ❌

File: qa/L0_parameters/parameters_test.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 
-# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright 2023-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (rmccorm4) [inline qa/L0_parameters/parameters_test.py:3]: Did you run a CI pipeline?

File: src/command_line_parser.cc [hunk 5]
```diff
@@ -1345,7 +1350,8 @@ TritonParser::Parse(int argc, char** argv)
           lparams.http_address_ = optarg;
           break;
         case OPTION_HTTP_HEADER_FORWARD_PATTERN:
-          lparams.http_forward_header_pattern_ = optarg;
+          lparams.http_forward_header_pattern_ =
+              std::move(case_insensitive_prefix + optarg);
```
Comment (rmccorm4) [inline src/command_line_parser.cc:1354]: We should also update the docs on the header forward pattern feature to specifically mention that it will match case insensitively.

Comment (rmccorm4) [reply src/command_line_parser.cc:1354]: https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_parameters.md#forwarding-httpgrpc-headers-as-parameters

File: TRITON_VERSION [hunk 6]
```diff
@@ -1 +1 @@
-2.43.0dev
+2.44.0dev
```
Comment (GuanLuo) [inline TRITON_VERSION:1]: Why version is changed

File: docs/protocol/extension_statistics.md [hunk 7]
```diff
@@ -428,6 +464,26 @@ message InferStatistics
...
+  // The count and cumulative duration for successful responses.
+  StatisticDuration success = 3;
+
+  // The count and cumulative duration for failed responses.
+  StatisticDuration fail = 4;
+
+  // The count and cumulative duration for empty responses.
+  StatisticDuration empty_response = 5;
+}
+
```
Comment (GuanLuo) [inline docs/protocol/extension_statistics.md:486]: Clean up your branch so it doesn't include unrelated changes.

File: qa/L0_parameters/parameters_test.py [hunk 8]
```diff
@@ -54,13 +54,14 @@ async def asyncSetUp(self):
...
 
         self.parameter_list = []
-        self.parameter_list.append({"key1": "value1", "key2": "value2"})
-        self.parameter_list.append({"key1": 1, "key2": 2})
-        self.parameter_list.append({"key1": 123.123, "key2": 321.321})
-        self.parameter_list.append({"key1": True, "key2": "value2"})
-        self.parameter_list.append({"triton_": True, "key2": "value2"})
-
-        if TEST_HEADER == "1":
+        if TEST_HEADER in ["0", "1"]:
```
Comment (GuanLuo) [inline qa/L0_parameters/parameters_test.py:57]: I don't understand why `parameter_list` is constructed under condition, why it is not needed in case "2"?

File: qa/L0_response_statistics/test.sh [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+kill $SERVER_PID
+wait $SERVER_PID
+
+if [ $RET -eq 0 ]; then
+    echo -e "\n***\n*** Test Passed\n***"
+else
+    echo -e "\n***\n*** Test FAILED\n***"
+fi
+exit $RET
```
Comment (GuanLuo) [inline qa/L0_response_statistics/test.sh:88]: Remove unrelated changes.

File: qa/L0_parameters/parameters_test.py [hunk 10]
```diff
@@ -54,13 +54,14 @@ async def asyncSetUp(self):
...
+            self.parameter_list.append({"key1": "value1", "key2": "value2"})
+            self.parameter_list.append({"key1": 1, "key2": 2})
+            self.parameter_list.append({"key1": 123.123, "key2": 321.321})
+            self.parameter_list.append({"key1": True, "key2": "value2"})
+            self.parameter_list.append({"triton_": True, "key2": "value2"})
+
+        if TEST_HEADER in ["1", "2"]:
```
Comment (GuanLuo) [inline qa/L0_parameters/parameters_test.py:64]: Give meaningful value to `TEST_HEADER`, what does 0, 1, 2 imply?

File: src/grpc/grpc_server.cc [hunk 11]
```diff
@@ -1100,6 +1100,96 @@ CommonHandler::RegisterModelStatistics()
...
+              res.mutable_empty_response()->set_count(val);
+              err = stat_json.MemberAsUInt("ns", &val);
+              GOTO_IF_ERR(err, earlyexit);
+              res.mutable_empty_response()->set_ns(val);
+            }
+
+            (*statistics->mutable_response_stats())[key] = std::move(res);
+          }
+        }
+
```
Comment (GuanLuo) [inline src/grpc/grpc_server.cc:1192]: Remove

File: qa/L0_parameters/test.sh [hunk 12]
```diff
@@ -53,13 +53,22 @@ mkdir -p "${MODELDIR}/ensemble/1"
...
 
   # TEST_HEADER is a parameter used by `parameters_test.py` that controls
   # whether the script will test for inclusion of headers in parameters or not.
-  if [ $i == 1 ]; then
+  if [ $i == "test_parameters_and_headers" ]; then
     SERVER_ARGS="--model-repository=${MODELDIR} --exit-timeout-secs=120 --grpc-header-forward-pattern my_header.* --http-header-forward-pattern my_header.*"
+  elif [ $i == "test_header_forward_pattern_case_insensitive" ]; then
+    SERVER_ARGS="--model-repository=${MODELDIR} --exit-timeout-secs=120 --grpc-header-forward-pattern MY_HEADER.* --http-header-forward-pattern MY_HEADER.*"
+  elif [ $i == "test_grpc_header_forward_pattern_case_sensitive" ]; then
+    SERVER_ARGS="--model-repository=${MODELDIR} --exit-timeout-secs=120 --grpc-header-forward-pattern (?-i)MY_HEADER.*
```
Comment (rmccorm4) [inline qa/L0_parameters/test.sh:71]: this is missing a quote at the end, causing test failure
```suggestion
    SERVER_ARGS="--model-repository=${MODELDIR} --exit-timeout-secs=120 --grpc-header-forward-pattern (?-i)MY_HEADER.*"
```

Comment (yinggeh) [reply qa/L0_parameters/test.sh]: That's why we should keep each line short. Nice catch!

File: docs/protocol/extension_parameters.md [hunk 13]
```diff
@@ -98,6 +98,12 @@ value. For example to forward all the headers that start with 'PREFIX_' from
 both HTTP and GRPC, you should add `--http-header-forward-pattern PREFIX_.*
 --grpc-header-forward-pattern PREFIX_.*` to your `tritonserver` command.
 
+By default, the regular expression pattern matches headers with case-insensitive
+mode according to the HTTP protocol. If you want to enforce case-sensitive mode,
+simplying adding the `(?-i)` prefix which turns off case-insensitive mode, e.g.
+`--http-header-forward-pattern (?-i)PREFIX_.*`. Note `(?-i)` prefix is not supported
+on HTTP Python client.
```
Comment (rmccorm4) [inline docs/protocol/extension_parameters.md:105]:
```suggestion
`--http-header-forward-pattern (?-i)PREFIX_.*`. Note, headers sent through the
Python HTTP client may be automatically lower-cased by internal client libraries.
```

File: qa/L0_parameters/test.sh [hunk 14]
```diff
@@ -53,15 +53,24 @@ mkdir -p "${MODELDIR}/ensemble/1"
...
-  else
-    SERVER_ARGS="--model-repository=${MODELDIR} --exit-timeout-secs=120"
+  SERVER_ARGS="--model-repository=${MODELDIR} --exit-timeout-secs=120"
+  if [ "$i" == "test_parameters_and_headers" ]; then
+    SERVER_ARGS+=" --grpc-header-forward-pattern my_header.*"
+    SERVER_ARGS+=" --http-header-forward-pattern my_header.*"
+  elif [ "$i" == "test_header_forward_pattern_case_insensitive" ]; then
+    SERVER_ARGS+=" --grpc-header-forward-pattern MY_HEADER.*"
+    SERVER_ARGS+=" --http-header-forward-pattern MY_HEADER.*"
+  elif [ "$i" == "test_grpc_header_forward_pattern_case_sensitive" ]; then
```
Comment (rmccorm4) [inline qa/L0_parameters/test.sh:72]:
```suggestion
  # NOTE: headers sent through the python HTTP client may be automatically
  # lowercased by internal libraries like geventhttpclient, so we only test 
  # GRPC client for case-sensitivity here:
  # https://github.com/geventhttpclient/geventhttpclient/blob/d1e14356c3b02099c879cf9b3bdb684a0cbd8bf5/src/geventhttpclient/header.py#L62-L63
  elif [ "$i" == "test_grpc_header_forward_pattern_case_sensitive" ]; then
```

**Review by rmccorm4 — COMMENTED** [review]
LGTM other than minor doc/comment changes. Don't need to re-run pipeline again for these changes

File: qa/L0_parameters/test.sh (hunk 14 above)
Comment (GuanLuo) [reply qa/L0_parameters/test.sh:72]: I don't think the name of this case is in `all_tests` (`test_grpc_header_forward_pattern_case_sensitive`)

Comment (yinggeh) [reply qa/L0_parameters/test.sh:76]: Sorry that was my fault. Fixing it now.

File: docs/protocol/extension_parameters.md (hunk 13 above)
Comment (yinggeh) [reply docs/protocol/extension_parameters.md:105]: Updated.

Comment (rmccorm4) [reply docs/protocol/extension_parameters.md:105]: Where is the update? I don't see it

File: docs/protocol/extension_parameters.md [hunk 15]
```diff
@@ -1,5 +1,5 @@
 <!--
-# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright (c) 2023-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
[the remark this answers is not in the document]
Comment (yinggeh) [reply docs/protocol/extension_parameters.md:2]: Why removing (c)?

File: docs/protocol/extension_parameters.md (hunk 13 above)
Comment (yinggeh) [reply docs/protocol/extension_parameters.md]: Bruh. I did not include this change in the commit. Should be there now.

File: docs/protocol/extension_parameters.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (rmccorm4) [reply docs/protocol/extension_parameters.md]: It's a practice we adopted to keep line width < 80 characters where we can help it. It's not a problem in python/bash with the `#` comment, but for C++ the `//` goes to 81 characters with the `(c)`:

```
// 76 characters width
// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

// 81 characters width
// Copyright (c) 2023-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

// 77 characters with
// Copyright 2023-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

I believe we just followed this direction from another DLFW team internally

**Review by rmccorm4 — APPROVED** [review]

File: docs/protocol/extension_parameters.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (yinggeh) [reply docs/protocol/extension_parameters.md]: What about python/bash?

[the remark this answers is not in the document]
Comment (rmccorm4) [reply docs/protocol/extension_parameters.md]: Yes, we try to do it everywhere to be consistent
